cmd/charm/charmcmd/suite_test.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the GPLv3, see LICENCE file for details.
+
+package charmcmd_test
+
+import (
+	"reflect"
+	"strings"
+
+	qt "github.com/frankban/quicktest"
+)
+
+// TODO factor the RunSuite function into a separate public repo
+// somewhere.
+
+// RunSuite runs all methods on the given value that have the
+// prefix "Test". The signature of the test methods must be
+// func(*quicktest.C).
+//
+// If suite is is a pointer, the value pointed to is copied
+// before any methods are invoked on it; a new copy
+// is made for each test.
+//
+// If there is an Init(*quicktest.C) method, it will be
+// invoked before each test method runs.
+func RunSuite(c *qt.C, suite interface{}) {
+	sv := reflect.ValueOf(suite)
+	st := sv.Type()
+	for i := 0; i < st.NumMethod(); i++ {
+		methodName := st.Method(i).Name
+		name := strings.TrimPrefix(methodName, "Test")
+		if len(name) == len(methodName) {
+			continue
+		}
+		sv := sv
+		if st.Kind() == reflect.Ptr {
+			// Make a copy (this makes it possible to have
+			// parallel tests).
+			sv1 := reflect.New(st.Elem())
+			sv1.Elem().Set(sv.Elem())
+			sv = sv1
+		}
+		c.Run(name, func(c *qt.C) {
+			if s, ok := sv.Interface().(initSuite); ok {
```

Comment:
I changed my mind about using regular interface conversion here. If someone defines an Init method with the wrong signature, I think we want it to panic rather than just ignoring it. YMMV.